core/src/main/java/com/linecorp/armeria/internal/client/ClientPendingThrowableUtil.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.client;
+
+import static java.util.Objects.requireNonNull;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.Exceptions;
+
+import io.netty.util.AttributeKey;
+
+/**
+ * Sets a pending {@link Throwable} for the specified {@link ClientRequestContext}.
+ * This throwable will be thrown after all decorators are executed, but before the
+ * actual client execution starts.
+ *
+ * <p>For example:<pre>{@code
+ * final RuntimeException e = new RuntimeException();
+ * final WebClient webClient =
+ *         WebClient.builder(SessionProtocol.HTTP, endpointGroup)
+ *                  .contextCustomizer(ctx -> setUnprocessedPendingThrowable(ctx, e))
+ *                  .decorator(LoggingClient.newDecorator()) // the request is logged
+ *                  .build();
+ * assertThatThrownBy(() -> webClient.blocking().get("/"))
+ *         .isInstanceOf(UnprocessedRequestException.class)
+ *         .hasCause(e);
+ * }</pre>
+ */
+public final class ClientPendingThrowableUtil {
+
+    private static final AttributeKey<PendingThrowableContainer> CLIENT_PENDING_THROWABLE =
+            AttributeKey.valueOf(ClientPendingThrowableUtil.class, "CLIENT_PENDING_THROWABLE");
+
+    /**
+     * Sets a pending {@link Throwable} for the specified {@link ClientRequestContext}.
+     * Note that the set throwable will be peeled via {@link Exceptions#peel(Throwable)}
+     * before being set.
+     *
+     * <p>For example:<pre>{@code
+     * final RuntimeException e = new RuntimeException();
+     * final CompletionException wrapper = new CompletionException(e);
+     * final ClientRequestContext ctx = ...
+     * setPendingThrowable(ctx, wrapper);
+     * final Throwable throwable = pendingThrowable(ctx);
+     * assert throwable != null;
+     * assertThat(throwable).isEqualTo(e);
+     * }</pre>
+     */
+    public static void setPendingThrowable(ClientRequestContext ctx, Throwable cause) {
+        requireNonNull(ctx, "ctx");
+        requireNonNull(cause, "cause");
+        cause = Exceptions.peel(cause);
+        final PendingThrowableContainer container = new PendingThrowableContainer(cause, ctx);
```

Comment:
Could you explain why we need the container instead of storing the `Throwable` directly? 🤔

Reply:
My original intention was to avoid a breaking change.

To start with, at first I wanted `PENDING_THROWABLE` to not be copied to derived attributes by default. (this is the reason I defined a `PendingThrowableContainer` and compared context referential equality)
The reason for this was I was worried it might be a breaking change for users who had been using `ClientRequestContext#newDerivedContext` in decorators.

For instance, say that a user implements a custom decorator which derives the context, and sets an endpoint manually. With the current change, all requests using the derived context would fail if the `PENDING_THROWABLE` is set in the original context.

Having said this, I'm unsure if there are such use-cases, so I've reverted to `PENDING_THROWABLE` being inherited for now.